### PR TITLE
Refactor dataset diff and compute metric

### PR DESCRIPTION
**Pull request by changhiskhan**
1. Minor bug fixes
2. Add `ds.compute_metric(func)` and `ds.diff(v1, v2)` to dataset interface

### CODE REVIEW

File: python/lance/_lib.pyx [hunk 1]
```diff
@@ -363,6 +384,34 @@ cdef class FileSystemDataset(Dataset):
...
+        return LanceDiff(self.checkout(v1), self.checkout(v2))
+
+    def compute_metric(self, metric_func: Callable[[Dataset], pd.DataFrame],
+                       versions: list[int] = None) -> pd.DataFrame:
+        """
+        Compute a metric across versions for this dataset
+
+        Parameters
+        ----------
+        metric_func: FileSystemDataset -> pd.DataFrame
```
Comment (eddyxu) [inline python/lance/_lib.pyx:407]: `Callalble[[FileSystmeDataset], pd.DataFrame]`

File: python/lance/_lib.pyx [hunk 2]
```diff
@@ -363,6 +384,34 @@ cdef class FileSystemDataset(Dataset):
...
+            Start version
+        v2: int
+            End version
+        """
+        return LanceDiff(self.checkout(v1), self.checkout(v2))
+
+    def compute_metric(self, metric_func: Callable[[Dataset], pd.DataFrame],
+                       versions: list[int] = None) -> pd.DataFrame:
+        """
+        Compute a metric across versions for this dataset
```
Comment (eddyxu) [inline python/lance/_lib.pyx:403]: Main concerns are the stability / maturity of these two APIs.

File: python/lance/_lib.pyx [hunk 3]
```diff
@@ -363,6 +384,34 @@ cdef class FileSystemDataset(Dataset):
...
+
+        Parameters
+        ----------
+        metric_func: FileSystemDataset -> pd.DataFrame
+            Function to compute some arbitrary metric for a given version
+        versions: list of int, default None
+            Compute for specified versions. Compute for all versions if None.
```
Comment (eddyxu) [inline python/lance/_lib.pyx:410]: Should the default behavior to just compute metrics for `latest version` ?

Comment (changhiskhan) [reply python/lance/_lib.pyx:410]: i don't think you'd really need this method then. If you just wanted to compute on latest version, `metric_func(ds)` is simpler than `compute_metrics(ds, metric_func)`. wdyt?

File: python/lance/_lib.pyx (hunk 2 above)
Comment (changhiskhan) [reply python/lance/_lib.pyx:403]: as discussed, moved to `lance/__init__.py` and not as a Dataset method

File: python/lance/_lib.pyx (hunk 1 above)
Comment (changhiskhan) [reply python/lance/_lib.pyx:407]: fixed

**Review by eddyxu — APPROVED** [review]
LGTM

**Pull request merged**